activerecord/lib/active_record/transactions.rb

```diff
@@ -350,9 +350,9 @@ def rolledback!(force_restore_state: false, should_run_callbacks: true) #:nodoc:
     end
 
     # Add the record to the current transaction so that the #after_rollback and #after_commit callbacks
-    # can be called.
+    # can be called and state restored.
     def add_to_transaction
-      if has_transactional_callbacks?
+      if has_transactional_callbacks? || has_remembered_start_transaction_state?
```

Comment:
I agree that the condition is not at the best location, but something like this is needed for the record to be added to the transaction's list that is [checked when a rollback happens](https://github.com/rails/rails/blob/c741b493409a5d8ae1f2514f73570a014ddc6b57/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L64).